src/tests/end2end/D3D12ResidencyManagerTests.cpp

```diff
@@ -279,12 +279,19 @@ TEST_F(D3D12ResidencyManagerTests, OverBudgetUsingBudgetNotifications) {
 
     const D3D12_RESOURCE_DESC bufferDesc = CreateBasicBufferDesc(GPGMM_MB_TO_BYTES(1));
 
+    ALLOCATION_DESC bufferAllocationDesc = {};
+    bufferAllocationDesc.HeapType = D3D12_HEAP_TYPE_DEFAULT;
+
+    const DXGI_MEMORY_SEGMENT_GROUP bufferMemorySegment =
+        residencyManager->GetMemorySegmentGroup(bufferAllocationDesc.HeapType);
+    const uint64_t memoryUnderBudget = GetBudgetLeft(residencyManager.Get(), bufferMemorySegment);
+
     // Keep allocating until we reach the budget.
     std::vector<ComPtr<ResourceAllocation>> allocations = {};
-    while (!IsOverBudget(residencyManager.Get())) {
+    while (resourceAllocator->GetInfo().UsedMemoryUsage < memoryUnderBudget) {
```

Comment:
The suggestion is good but the loop should still terminate since the allocated amount always increases. However, it's not guarenteed that Evict won't be called, where creating the last heap, exceeds budget, making the first not resident. So the condition you've suggested is still needed but not enough - must also disable heap growth.